src/lightning/data/streaming/dataset.py

```diff
@@ -85,23 +72,56 @@ def __init__(
         self.index = 0
         self.has_triggered_download = False
         self.min_items_per_replica: Optional[int] = None
-        self.seed = seed
         self.current_epoch = 0
         self.random_state = None
+        self.shuffler: Optional[Shuffle] = None
+
+    def _setup(self) -> None:
+        if self.worker_env is None:
+            self.worker_env = _WorkerEnv.detect()
+
+        env = Environment(dist_env=self.distributed_env, worker_env=self.worker_env)
+
+        # TODO: Why are we conflating input dir and cache dir into one thing? They are conceptually different!
+        # Override the provided input_path
+        cache_dir = _try_create_cache_dir(shard_rank=env.shard_rank)
+        if cache_dir:
+            self.input_dir.path = cache_dir
+
+        self.cache = Cache(input_dir=self.input_dir, item_loader=self.item_loader, chunk_bytes=1)
+        self.cache._reader._try_load_config()
+
+        if not self.cache.filled:
+            raise ValueError(
+                f"The provided dataset `{self.input_dir}` doesn't contain any {_INDEX_FILENAME} file."
+                " HINT: Did you successfully optimize a dataset to the provided `input_dir`?"
+            )
+
+        self.shuffler = (
+            FullShuffle(self.cache, self.seed, self.drop_last)
+            if self.shuffle
+            else NoShuffle(self.cache, self.seed, self.drop_last)
+        )
 
     def __len__(self) -> int:
-        return self.shuffle.get_len(self.distributed_env, self.current_epoch)
+        if self.cache is None:
+            self._setup()
```

Comment:
If you call `length` and then use `DataLoader`, do you get different `cache_dir` for the workers or the Cache object is pickled over ?

Reply:
I made sure that we recreate the cache if needed. But I think given your question, I need to add a test to assert these side effects when calling iter(), getitem(), and len()